File: frontend/cls-budget-web/src/features/pay-schedules/utils/payDateCalculator.ts
import {
  PayFrequencyTypeIds,
  type PayScheduleConfig,
  type PayScheduleResponse,
} from "@/features/pay-schedules/types/paySchedule";

export const END_OF_MONTH_DAY = 31;

export interface PayPeriodBoundary {
  periodStart: string;
  periodEnd: string;
  label: string;
}

export function toScheduleConfig(
  schedule: PayScheduleResponse,
): PayScheduleConfig {
  return {
    payFrequencyTypeId: schedule.payFrequencyTypeId,
    anchorDate: schedule.anchorDate,
    semiMonthlyDay1: schedule.semiMonthlyDay1,
    semiMonthlyDay2: schedule.semiMonthlyDay2,
  };
}

export function toUtcDate(iso: string): Date {
  const value = iso.includes("T") ? iso : `${iso}T00:00:00.000Z`;
  const date = new Date(value);
  return new Date(
    Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate()),
  );
}

export function toUtcDateIso(date: Date): string {
  return new Date(
    Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate()),
  ).toISOString();
}

export function resolveSemiMonthlyDay(
  year: number,
  month: number,
  day: number,
): Date {
  const lastDay = new Date(Date.UTC(year, month + 1, 0)).getUTCDate();
  const resolvedDay = day === END_OF_MONTH_DAY ? lastDay : Math.min(day, lastDay);
  return new Date(Date.UTC(year, month, resolvedDay));
}

export function getPayDatesInRange(
  schedule: PayScheduleConfig,
  rangeStart: string,
  rangeEnd: string,
): string[] {
  const start = toUtcDate(rangeStart);
  const end = toUtcDate(rangeEnd);
  if (end.getTime() < start.getTime()) return [];

  switch (schedule.payFrequencyTypeId) {
    case PayFrequencyTypeIds.Weekly:
      return getIntervalPayDates(start, end, schedule.anchorDate, 7);
    case PayFrequencyTypeIds.BiWeekly:
      return getIntervalPayDates(start, end, schedule.anchorDate, 14);
    case PayFrequencyTypeIds.SemiMonthly:
      return getSemiMonthlyPayDates(
        start,
        end,
        schedule.semiMonthlyDay1,
        schedule.semiMonthlyDay2,
      );
    default:
      return [];
  }
}

export function getPayPeriodBoundaries(
  schedule: PayScheduleConfig,
  rangeStart: string,
  rangeEnd: string,
): PayPeriodBoundary[] {
  if (schedule.payFrequencyTypeId === PayFrequencyTypeIds.SemiMonthly) {
    return getSemiMonthlyHalfMonthBoundaries(schedule, rangeStart, rangeEnd);
  }

  const start = toUtcDate(rangeStart);
  const end = toUtcDate(rangeEnd);
  const payDates = getPayDatesInRange(schedule, rangeStart, rangeEnd).map(
    toUtcDate,
  );

  if (payDates.length === 0) {
    return [
      {
        periodStart: toUtcDateIso(start),
        periodEnd: toUtcDateIso(end),
        label: `${formatShortDate(start)} – ${formatShortDate(end)}`,
      },
    ];
  }

  const boundaries: PayPeriodBoundary[] = [];
  const lastPayBeforeStart = getLastPayDateBeforeRange(schedule, rangeStart);
  let previousExclusive = lastPayBeforeStart
    ? toUtcDate(lastPayBeforeStart)
    : addDays(start, -1);

  for (const payDate of payDates) {
    const periodStart = addDays(previousExclusive, 1);
    boundaries.push({
      periodStart: toUtcDateIso(periodStart),
      periodEnd: toUtcDateIso(payDate),
      label: `Through ${formatShortDate(payDate)}`,
    });
    previousExclusive = payDate;
  }

  if (previousExclusive.getTime() < end.getTime()) {
    const periodStart = addDays(previousExclusive, 1);
    boundaries.push({
      periodStart: toUtcDateIso(periodStart),
      periodEnd: toUtcDateIso(end),
      label: `After ${formatShortDate(periodStart)}`,
    });
  }

  return boundaries;
}

function getSemiMonthlyHalfMonthBoundaries(
  schedule: PayScheduleConfig,
  rangeStart: string,
  rangeEnd: string,
): PayPeriodBoundary[] {
  const start = toUtcDate(rangeStart);
  const end = toUtcDate(rangeEnd);
  const day1 = schedule.semiMonthlyDay1;
  const day2 = schedule.semiMonthlyDay2;

  if (day1 == null || day2 == null) {
    return [
      {
        periodStart: toUtcDateIso(start),
        periodEnd: toUtcDateIso(end),
        label: `${formatShortDate(start)} – ${formatShortDate(end)}`,
      },
    ];
  }

  const splitDay = Math.max(day1, day2);
  const splitDate = resolveSemiMonthlyDay(
    start.getUTCFullYear(),
    start.getUTCMonth(),
    splitDay,
  );
  const splitLabel = formatShortDate(splitDate);
  const period1End =
    splitDate.getTime() <= end.getTime() ? splitDate : end;
  const period2Start = addDays(splitDate, 1);

  if (period2Start.getTime() > end.getTime()) {
    return [
      {
        periodStart: toUtcDateIso(start),
        periodEnd: toUtcDateIso(end),
        label: `Before ${splitLabel}`,
      },
    ];
  }

  return [
    {
      periodStart: toUtcDateIso(start),
      periodEnd: toUtcDateIso(period1End),
      label: `Before ${splitLabel}`,
    },
    {
      periodStart: toUtcDateIso(period2Start),
      periodEnd: toUtcDateIso(end),
      label: `After ${splitLabel}`,
    },
  ];
}

/** Paychecks before the budget month start belong in the first pay period. */
export function resolvePaymentPayPeriodIndex(
  paymentDate: string,
  periods: ReadonlyArray<{ periodStart: string; periodEnd: string }>,
  budgetStartPeriod: string,
): number {
  const periodIndex = periods.findIndex((period) =>
    paymentDateInPeriod(paymentDate, period.periodStart, period.periodEnd),
  );
  if (periodIndex >= 0) return periodIndex;
  if (periods.length === 0) return -1;

  const paymentMs = toUtcDate(paymentDate).getTime();
  const budgetStartMs = toUtcDate(budgetStartPeriod).getTime();
  const firstPeriodStartMs = toUtcDate(periods[0].periodStart).getTime();

  if (paymentMs < budgetStartMs || paymentMs < firstPeriodStartMs) {
    return 0;
  }

  return -1;
}

function getLastPayDateBeforeRange(
  schedule: PayScheduleConfig,
  rangeStart: string,
): string | null {
  const start = toUtcDate(rangeStart);
  const lookbackStart = addDays(start, -35);
  const lookbackEnd = addDays(start, -1);
  const dates = getPayDatesInRange(
    schedule,
    toUtcDateIso(lookbackStart),
    toUtcDateIso(lookbackEnd),
  );
  return dates.length > 0 ? dates[dates.length - 1]! : null;
}

function getIntervalPayDates(
  rangeStart: Date,
  rangeEnd: Date,
  anchorDate: string | null,
  intervalDays: number,
): string[] {
  if (!anchorDate) return [];

  let cursor = toUtcDate(anchorDate);
  while (cursor.getTime() > rangeStart.getTime()) {
    cursor = addDays(cursor, -intervalDays);
  }
  while (cursor.getTime() < rangeStart.getTime()) {
    cursor = addDays(cursor, intervalDays);
  }

  const dates: string[] = [];
  while (cursor.getTime() <= rangeEnd.getTime()) {
    dates.push(toUtcDateIso(cursor));
    cursor = addDays(cursor, intervalDays);
  }

  return dates;
}

function getSemiMonthlyPayDates(
  rangeStart: Date,
  rangeEnd: Date,
  day1: number | null,
  day2: number | null,
): string[] {
  if (
    day1 == null ||
    day2 == null ||
    day1 < 1 ||
    day1 > 31 ||
    day2 < 1 ||
    day2 > 31
  ) {
    return [];
  }

  const dates: string[] = [];
  let cursor = new Date(
    Date.UTC(rangeStart.getUTCFullYear(), rangeStart.getUTCMonth(), 1),
  );
  const endMonth = new Date(
    Date.UTC(rangeEnd.getUTCFullYear(), rangeEnd.getUTCMonth(), 1),
  );

  while (cursor.getTime() <= endMonth.getTime()) {
    for (const day of [...new Set([day1, day2])].sort((a, b) => a - b)) {
      const payDate = resolveSemiMonthlyDay(
        cursor.getUTCFullYear(),
        cursor.getUTCMonth(),
        day,
      );
      if (
        payDate.getTime() >= rangeStart.getTime() &&
        payDate.getTime() <= rangeEnd.getTime()
      ) {
        dates.push(toUtcDateIso(payDate));
      }
    }
    cursor = new Date(
      Date.UTC(cursor.getUTCFullYear(), cursor.getUTCMonth() + 1, 1),
    );
  }

  dates.sort();
  return dates;
}

function addDays(date: Date, days: number): Date {
  const next = new Date(date);
  next.setUTCDate(next.getUTCDate() + days);
  return next;
}

function formatShortDate(date: Date): string {
  return date.toLocaleDateString("en-US", {
    month: "numeric",
    day: "numeric",
    timeZone: "UTC",
  });
}

export function paymentDateInPeriod(
  paymentDate: string,
  periodStart: string,
  periodEnd: string,
): boolean {
  const paymentMs = toUtcDate(paymentDate).getTime();
  return (
    paymentMs >= toUtcDate(periodStart).getTime() &&
    paymentMs <= toUtcDate(periodEnd).getTime()
  );
}

File: backend/src/CLS.Budget.Application/PaySchedules/PayDateCalculator.cs
using CLS.Budget.Domain;
using CLS.Budget.Domain.Entities;

namespace CLS.Budget.Application.PaySchedules;

public static class PayDateCalculator
{
    public const int EndOfMonthDay = 31;

    public static IReadOnlyList<DateTime> GetPayDatesInRange(
        PaySchedule schedule,
        DateTime rangeStart,
        DateTime rangeEnd)
    {
        var start = ToUtcDate(rangeStart);
        var end = ToUtcDate(rangeEnd);
        if (end < start)
        {
            return [];
        }

        return schedule.PayFrequencyTypeId switch
        {
            PayFrequencyTypeIds.Weekly => GetIntervalPayDates(start, end, schedule.AnchorDate, 7),
            PayFrequencyTypeIds.BiWeekly => GetIntervalPayDates(start, end, schedule.AnchorDate, 14),
            PayFrequencyTypeIds.SemiMonthly => GetSemiMonthlyPayDates(
                start,
                end,
                schedule.SemiMonthlyDay1,
                schedule.SemiMonthlyDay2),
            _ => throw new ArgumentOutOfRangeException(
                nameof(schedule.PayFrequencyTypeId),
                schedule.PayFrequencyTypeId,
                "Unsupported pay frequency type.")
        };
    }

    public static IReadOnlyList<PayPeriodBoundary> GetPayPeriodBoundaries(
        PaySchedule schedule,
        DateTime rangeStart,
        DateTime rangeEnd)
    {
        var start = ToUtcDate(rangeStart);
        var end = ToUtcDate(rangeEnd);

        if (schedule.PayFrequencyTypeId == PayFrequencyTypeIds.SemiMonthly)
        {
            return GetSemiMonthlyHalfMonthBoundaries(schedule, start, end);
        }

        var payDates = GetPayDatesInRange(schedule, start, end);

        if (payDates.Count == 0)
        {
            return
            [
                new PayPeriodBoundary(start, end, $"{start:M/d} – {end:M/d}")
            ];
        }

        var boundaries = new List<PayPeriodBoundary>();
        var lastPayBeforeStart = GetLastPayDateBeforeRange(schedule, start);
        var previousExclusive = lastPayBeforeStart ?? start.AddDays(-1);

        foreach (var payDate in payDates)
        {
            boundaries.Add(new PayPeriodBoundary(
                previousExclusive.AddDays(1),
                payDate,
                FormatPeriodLabel(previousExclusive.AddDays(1), payDate, payDate)));
            previousExclusive = payDate;
        }

        if (previousExclusive < end)
        {
            boundaries.Add(new PayPeriodBoundary(
                previousExclusive.AddDays(1),
                end,
                FormatPeriodLabel(previousExclusive.AddDays(1), end, null, isAfter: true)));
        }

        return boundaries;
    }

    private static IReadOnlyList<PayPeriodBoundary> GetSemiMonthlyHalfMonthBoundaries(
        PaySchedule schedule,
        DateTime start,
        DateTime end)
    {
        if (schedule.SemiMonthlyDay1 is null or < 1 or > 31 ||
            schedule.SemiMonthlyDay2 is null or < 1 or > 31)
        {
            return [new PayPeriodBoundary(start, end, $"{start:M/d} – {end:M/d}")];
        }

        var splitDay = Math.Max(schedule.SemiMonthlyDay1.Value, schedule.SemiMonthlyDay2.Value);
        var splitDate = ResolveSemiMonthlyDay(start.Year, start.Month, splitDay);
        var period1End = splitDate <= end ? splitDate.AddDays(-1) : end;
        var period2Start = splitDate;

        if (period2Start > end)
        {
            return
            [
                new PayPeriodBoundary(start, end, $"Before {splitDate:M/d}")
            ];
        }

        if (period1End < start)
        {
            return
            [
                new PayPeriodBoundary(period2Start, end, $"After {splitDate:M/d}")
            ];
        }

        return
        [
            new PayPeriodBoundary(start, period1End, $"Before {splitDate:M/d}"),
            new PayPeriodBoundary(period2Start, end, $"After {splitDate:M/d}")
        ];
    }

    internal static DateTime ToUtcDate(DateTime value) =>
        DateTime.SpecifyKind(value.Date, DateTimeKind.Utc);

    internal static DateTime ResolveSemiMonthlyDay(int year, int month, int day)
    {
        var lastDay = DateTime.DaysInMonth(year, month);
        var resolvedDay = day == EndOfMonthDay ? lastDay : Math.Min(day, lastDay);
        return new DateTime(year, month, resolvedDay, 0, 0, 0, DateTimeKind.Utc);
    }

    private static IReadOnlyList<DateTime> GetIntervalPayDates(
        DateTime rangeStart,
        DateTime rangeEnd,
        DateTime? anchorDate,
        int intervalDays)
    {
        if (anchorDate is null)
        {
            return [];
        }

        var anchor = ToUtcDate(anchorDate.Value);
        var cursor = anchor;

        while (cursor > rangeStart)
        {
            cursor = cursor.AddDays(-intervalDays);
        }

        while (cursor < rangeStart)
        {
            cursor = cursor.AddDays(intervalDays);
        }

        var dates = new List<DateTime>();
        while (cursor <= rangeEnd)
        {
            dates.Add(cursor);
            cursor = cursor.AddDays(intervalDays);
        }

        return dates;
    }

    private static IReadOnlyList<DateTime> GetSemiMonthlyPayDates(
        DateTime rangeStart,
        DateTime rangeEnd,
        int? day1,
        int? day2)
    {
        if (day1 is null or < 1 or > 31 || day2 is null or < 1 or > 31)
        {
            return [];
        }

        var dates = new List<DateTime>();
        var cursor = new DateTime(rangeStart.Year, rangeStart.Month, 1, 0, 0, 0, DateTimeKind.Utc);
        var endMonth = new DateTime(rangeEnd.Year, rangeEnd.Month, 1, 0, 0, 0, DateTimeKind.Utc);

        while (cursor <= endMonth)
        {
            foreach (var day in new[] { day1.Value, day2.Value }.Distinct().Order())
            {
                var payDate = ResolveSemiMonthlyDay(cursor.Year, cursor.Month, day);
                if (payDate >= rangeStart && payDate <= rangeEnd)
                {
                    dates.Add(payDate);
                }
            }

            cursor = cursor.AddMonths(1);
        }

        dates.Sort();
        return dates;
    }

    private static DateTime? GetLastPayDateBeforeRange(PaySchedule schedule, DateTime rangeStart)
    {
        var start = ToUtcDate(rangeStart);
        var lookbackStart = start.AddDays(-35);
        var lookbackEnd = start.AddDays(-1);
        var dates = GetPayDatesInRange(schedule, lookbackStart, lookbackEnd);
        return dates.Count > 0 ? dates[^1] : null;
    }

    private static string FormatPeriodLabel(
        DateTime periodStart,
        DateTime periodEnd,
        DateTime? payDate,
        bool isAfter = false)
    {
        if (isAfter && payDate is null)
        {
            return $"After {periodStart:M/d}";
        }

        if (payDate is not null && periodStart == payDate && periodEnd == payDate)
        {
            return $"Through {payDate:M/d}";
        }

        if (payDate is not null)
        {
            return $"Through {payDate:M/d}";
        }

        return $"{periodStart:M/d} – {periodEnd:M/d}";
    }
}

public sealed record PayPeriodBoundary(
    DateTime PeriodStart,
    DateTime PeriodEnd,
    string Label);
